compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -476,12 +480,33 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
                 }
             }
             ProjectionElem::Downcast(name, index) => ProjectionElem::Downcast(name, index),
-            ProjectionElem::Field(f, ty) => ProjectionElem::Field(f, ty),
+            ProjectionElem::Field(f, ty) => {
+                if let Value::Aggregate(_, _, fields) = self.get(value) {
+                    return Some(fields[f.as_usize()]);
+                } else if let Value::Projection(outer_value, ProjectionElem::Downcast(_, read_variant)) = self.get(value)
+                    && let Value::Aggregate(_, written_variant, fields) = self.get(*outer_value)
+                    && written_variant == read_variant
```

Comment:
can this be violated? Could we just assert that they are equal?

Reply:
This cannot happen from surface rust, as MIR is built such as a switch on the discriminant dominates the downcasts.
Reading reference chapter on enums and RFC 2195, I understand that accessing the wrong variant is not UB if the enum has `repr`. So it's not impossible for a series of MIR opts to generate a downcast to an inactive variant. IIUC, miri does not check the variant either.
This can definitely happen from custom MIR. I'd rather silently fail to optimize (which is safe) than ICE.

Reply:
> This cannot happen from surface rust, as MIR is built such as a switch on the discriminant dominates the downcasts.
> Reading reference chapter on enums and RFC 2195, I understand that accessing the wrong variant is not UB if the enum has `repr`. So it's not impossible for a series of MIR opts to generate a downcast to an inactive variant.

this could be copied verbatim as a comment on the `==` expression

Reply:
More pragmatically, this pass is not aware of control-flow. We could have:
```rust
_1 = None
_2 = discriminant(_1)
SwitchInt(_2) -> [0: bb1, otherwise: bb2]

bb1:
// stuff

bb2:
_3 = (_1 as Some).0
// stuff
```

We must not ICE for the assignment to `_3`.